src/Network/Riak/Connection/Internal.hs

```diff
@@ -337,6 +346,124 @@ pipeline_ conn@Connection{..} reqs = do
   _ <- onIOException "pipeline_" $ waitBoth sendReqs receiveResps
   return ()
 
+data WorkerResult
+  = WorkerDone               -- Background request-sender has run out of requests.
+  | WorkerDied SomeException -- Background request-sender died somehow.
+
+stream
+  :: forall req resp.
+     Request req
+  => (Connection -> IO resp)
+  -> Connection
+  -> Pipes.Producer req IO ()
+  -> Pipes.Producer' (req, resp) IO ()
+stream receive conn@Connection{..} reqs = do
+  -- Keep track of the requests sent. This allows us to pair requests with their
+  -- corresponding responses.
+  requestsChan :: TChan req <-
+    lift newTChanIO
+
+  resultVar :: TMVar WorkerResult <-
+    lift newEmptyTMVarIO
+
+  -- Keep a weak reference to the worker result var: if the worker sees this var
+  -- has been garbage collected, it knows the driver of the returned producer
+  -- has died, or dropped the producer on the floor - point is, there's no one
+  -- left that cares about the responses, so don't bother sending any more
+  -- requests.
+  weakResultVar :: Weak (TMVar WorkerResult) <-
+    lift (mkWeakTMVar resultVar (pure ()))
+
+  -- Background thread: send all of the requests (or die trying).
+  let worker :: IO ()
+      worker = Pipes.runEffect (reqs >-> doSend)
+        where
+          doSend :: Pipes.Consumer' req IO ()
+          doSend = do
+            req <- Pipes.await
+            alive <- lift (isJust <$> deRefWeak weakResultVar)
+            when alive $ do
+              lift (sendRequest conn req)
+              lift (atomically (writeTChan requestsChan req))
+              doSend
+
+  -- Spawn the send thread; if it successfully completes, write WorkerDone.
+  -- Else, write the exception that brought it down.
+  --
+  -- 'mask' is used here to guarantee that:
+  --
+  -- 1. We start the worker action (thread doesn't die right after 'forkIO')
+  -- 2. We write WorkerDone after (thread doesn't die right after 'worker')
+  --
+  -- Use 'weakResultVar' instead of 'resultVar' so that only the driver keeps
+  -- 'resultVar' alive.
+  _ <-
+    lift $
+      mask $ \restore ->
+        forkIO $ do
```

Comment:
I think you should consider using `async` for this. @tmcgilchrist Thoughts, this looks like an implementation of what it will give you out of the box, from looking at the code it looks ok at the moment, but it would be nice not to have to maintain it in here.

Reply:
Could you clarify which part(s) of `async` you mean?

Here's a bit of info about the code is designed like this:

1. I started with the preferred output type `Producer (req, resp) IO ()`. This is the most natural type to consume streaming responses.
2. With this type, there is no way to acquire a "resource" like a background worker thread, file handle, etc - to be reaped after the caller is done with the stream.
3. Since we want to interleave making requests with consuming responses, a background worker is necessary (this is how `pipeline` works as well).
4. Since we can't scope the lifetime of the worker thread to the returned `Producer` without dealing in a `ResourceT IO`-type layer (ugly, complicated), we instead `forkIO` a worker and throw away its `ThreadId`. It's just sparked off and lost forever.
5. The returned `Producer` checks a `TMVar` to ascertain whether or not there might be more requests to be made (and thus it should wait around to `receive` the response). This `TMVar` is also the mechanism by which the background thread's lifetime is controlled - if the returned `Producer` is garbage collected, no one it left to consume responses, and the background worker will notice this, stop sending requests, and die.